animation_nodes/data_structures/meshes/mesh_data.pyx

```diff
@@ -370,3 +384,272 @@ def calculateLinkedVertices(int verticesAmount, EdgeIndicesList edges):
+    for i in range(polygonAmount):
+        polyLength = _oldPolyLengths[i]
+        if polyLength > 3:
+            triAmount += polyLength - 2
+        else:
+            triAmount += 1
+
+    cdef unsigned int *_oldIndices = polygons.indices.data
+    cdef unsigned int *_oldPolyStarts = polygons.polyStarts.data
+
+    cdef PolygonIndicesList newPolygons = PolygonIndicesList(
+                                          indicesAmount = 3 * triAmount,
+                                          polygonAmount = triAmount)
+    cdef unsigned int *_newIndices = newPolygons.indices.data
+    cdef unsigned int *_newPolyStarts = newPolygons.polyStarts.data
+    cdef unsigned int *_newPolyLengths = newPolygons.polyLengths.data
+
+    cdef Vector3 preVertex, earVertex, nexVertex, ab, bc, v1, v2, v3
+    cdef LongList neighbors, indices, angles, sortAngles
+    cdef Vector3DList polyVertices
+    cdef float angle
+    cdef Py_ssize_t polyStart, triIndex, polyIndex
+    cdef Py_ssize_t j, k, amount, index, preIndex, earIndex, nexIndex
+
+    triIndex = 0
+    polyIndex = 0
+    for i in range(polygonAmount):
+        polyLength = _oldPolyLengths[i]
+        polyStart = _oldPolyStarts[i]
+        neighbors = LongList(length = polyLength)
+        indices = LongList(length = polyLength)
+        for j in range(polyLength):
+            neighbors.data[j] = _oldIndices[polyStart + j]
+            indices.data[j] = j
+
+        polyVertices = transformVerticesOfPolygon(oldVertices, neighbors)
+        if not polyCounterClockwise(polyVertices):
+            neighbors = neighbors.reversed()
+            polyVertices = polyVertices.reversed()
+
+        for j in range(polyLength):
+            amount = indices.length
+
+            if amount == 3:
+                _newPolyStarts[triIndex] = polyIndex
+                _newPolyLengths[triIndex] = 3
+                triIndex += 1
+
+                _newIndices[polyIndex] = neighbors.data[indices.data[0]]
+                _newIndices[polyIndex + 1] = neighbors.data[indices.data[1]]
+                _newIndices[polyIndex + 2] = neighbors.data[indices.data[2]]
+                polyIndex += 3
+
+                break
+
+            # Calculate inner-angle for all vertices of polygon
+            angles = LongList(length = amount)
+            for k in range(amount):
+                preVertex = polyVertices.data[indices.data[(amount + k - 1) % amount]]
+                earVertex = polyVertices.data[indices.data[k]]
+                nexVertex = polyVertices.data[indices.data[(k + 1) % amount]]
+
+                subVec3(&ab, &earVertex, &preVertex)
+                subVec3(&bc, &earVertex, &nexVertex)
+
+                normalizeVec3_InPlace(&ab)
+                normalizeVec3_InPlace(&bc)
+
+                angle = angleNormalizedVec3(&ab, &bc)
+                angles.data[k] = int((angle - angle % 0.001) * 180.0 / pi)
+            sortAngles = LongList.fromNumpyArray(numpy.sort(angles.asMemoryView()).astype(int))
```

Comment:
Do we really need to recalculate everything everytime? I assume only two angles will change here. Also, using numpy here would be questionable. If you need to, use `qsort` from stdlib.

Reply:
> Do we really need to recalculate everything everytime? I assume only two angles will change here.

That can be avoided.

> Also, using numpy here would be questionable. If you need to, use `qsort` from stdlib.
 
Okay. I did not know that.

Reply:
Omar, I tried this but failed. Please help me,
```
from libc.stdlib cimport qsort

newAngles = LongList()
qsort(newAngles, newAngles.length, sizeof(int), compare)

cdef compare(const void * a, const void * b):
   return <int*>a - <int*>b
```

Reply:
`<int*>a` will cast `a` to an integer pointer, so you still need to dereference it to get the actual value `(<int*>a)[0]`.

Reply:
> Do we really need to recalculate everything everytime? I assume only two angles will change here. 

I have modified the algorithm, now no need to recalculate everything everytime.

Reply:
> `<int*>a` will cast `a` to an integer pointer, so you still need to dereference it to get the actual value `(<int*>a)[0]`.

I have done a test, but it is not working.
```
from libc.stdlib cimport qsort

qsort(newAngles, newAngles.length, sizeof(int), compare)

cdef compare(const void * a, const void * b):
   return (<int*>a)[0] - (<int*>b)[0]
```
![Screenshot from 2020-04-23 00-41-33](https://user-images.githubusercontent.com/30294746/80023491-3a6b7500-84fb-11ea-8769-c51e51713262.png)

Reply:
Looks like you need a pointer to the function. So try `qsort(newAngles, newAngles.length, sizeof(int), &compare)`.

Reply:
Sorry for late reply. It gave the same error, `Cannot assign type 'object (*)(const void *, const void *)' to 'int (*)(const void *, const void *) nogil'`

Reply:
We forgot the function type:
```
cdef int compare(const void * a, const void * b):
```

Reply:
Still the error is there,
![Screenshot from 2020-04-23 02-23-37](https://user-images.githubusercontent.com/30294746/80032924-832a2a80-8509-11ea-94ef-8ff9112f4751.png)

Reply:
Is this related, https://stackoverflow.com/a/50777451

Reply:
This code is working, 
```
from cpython cimport array
import array

cdef quickSort(LongList newAngles):
    cdef Py_ssize_t amount = newAngles.length
    cdef array.array b = array.array('i', newAngles.asNumpyArray())
    cdef int[:] a = b
    qsort(&a[0], amount, sizeof(int), &compare)
    print(b)

cdef int compare(const void * a, const void * b) with gil:
   return (<int*>a)[0] - (<int*>b)[0]
```

Reply:
You need to release the GIL for the compare function. See https://cython.readthedocs.io/en/latest/src/userguide/external_C_code.html#declaring-a-function-as-callable-without-the-gil

Reply:
Okay, but is it correct way to do that?

Reply:
It is in the signature of the expected callback, so I guess so.

Reply:
Okay, 
```
cdef quickSort(LongList newAngles):
    cdef Py_ssize_t amount = newAngles.length
    cdef array.array b = array.array('i', newAngles.asNumpyArray())
    cdef int[:] a = b
    with nogil:
        qsort(&a[0], amount, sizeof(int), &compare)
    print(b)

cdef int compare(const void * a, const void * b) with gil:
   return (<int*>a)[0] - (<int*>b)[0]
```

Reply:
This is a complete function,
```
cdef quickSort(LongList newAngles):
    cdef Py_ssize_t amount = newAngles.length
    cdef array.array b = array.array('i', newAngles.asNumpyArray())
    cdef int[:] a = b
    with nogil:
        qsort(&a[0], amount, sizeof(int), &compare)

    cdef LongList sortAngles = LongList(length = amount)
    cdef Py_ssize_t i
    for i in range(amount):
        sortAngles.data[i] = b[i]
    return sortAngles

cdef int compare(const void * a, const void * b) with gil:
   return (<int*>a)[0] - (<int*>b)[0]
```
Speed testing:
**With qsort:**
![test1_gil](https://user-images.githubusercontent.com/30294746/80095907-5831ec00-8586-11ea-8a85-5336f50b8cee.png)

**With Numpy:**
![test1_nogil](https://user-images.githubusercontent.com/30294746/80095942-6bdd5280-8586-11ea-8905-2cdcbafe0295.png)

Reply:
Can we use `sort` instead of `qsort` which is fast according this site: https://www.geeksforgeeks.org/c-qsort-vs-c-sort/?ref=rp

Reply:
I meant like this.

```cython
from libc.stdlib cimport qsort

cdef int compare(const void * a, const void * b) nogil:
   return (<int*>a)[0] - (<int*>b)[0]

qsort(newAngles, newAngles.length, sizeof(int), &compare)

```

Reply:
It gave an error, maybe need to `newAngles` to C-object,
![Screenshot from 2020-04-23 17-40-47](https://user-images.githubusercontent.com/30294746/80097905-aa284100-8589-11ea-8bed-e3effc057054.png)

Reply:
`newAngles` is a `LongList`. Use `newAngles.data`.

Reply:
Thank you so much for the help.

It is working But the output is not the same,
![Screenshot from 2020-04-23 17-50-38](https://user-images.githubusercontent.com/30294746/80098775-0f306680-858b-11ea-8bbf-567d0daa1487.png)

Reply:
A LongList is composed of longs, so I think `sizeof(int)` should be `sizeof(long)`.

Reply:
Ahhh, it worked. 

Thank you so much :)

Reply:
How is the performance now? Is it better than numpy? I am not sure if we are allowed to use c++ here, so not sure about the STL sort function.

Reply:
Yes, it is better.

But we use c++ in the Fast Noise.

Reply:
Based on your suggestions, now performance is far better. For example triangulation of a sphere, timing reduces to `~17ms` to `1.45ms`.

Reply:
Thanks. Okay.